File: src/API/app/domain/models/retirement_plan.py
from dataclasses import dataclass


@dataclass(slots=True)
class ContributionPlan:
    monthly_amount: float
    annual_increase_percent: float
    end_age: int


@dataclass(slots=True)
class ExpenseChange:
    age: int
    percent_change: float
    label: str | None = None


@dataclass(slots=True)
class SocialSecurityPlan:
    enabled: bool
    claim_age: int
    monthly_benefit: float
    annual_cola_percent: float


@dataclass(slots=True)
class RetirementPlan:
    id: str
    name: str
    current_age: int
    retirement_age: int
    longevity_age: int
    annual_roi_percent: float
    withdrawal_rate_percent: float
    annual_retirement_expense: float
    withdrawal_mode: str
    taxable_contribution: ContributionPlan
    retirement_contribution: ContributionPlan
    expense_changes: list[ExpenseChange]
    social_security: SocialSecurityPlan
    include_hsa_in_retirement: bool
    updated_at: str

File: src/UI/src/domain/retirementPlan.ts
export type ContributionPlan = {
  monthlyAmount: number;
  annualIncreasePercent: number;
  endAge: number;
};

export type ExpenseChange = {
  age: number;
  percentChange: number;
  label?: string | null;
};

export type SocialSecurityPlan = {
  enabled: boolean;
  claimAge: number;
  monthlyBenefit: number;
  annualColaPercent: number;
};

export type RetirementPlan = {
  id: string;
  name: string;
  currentAge: number;
  retirementAge: number;
  longevityAge: number;
  annualRoiPercent: number | null;
  annualRetirementExpense: number;
  taxableContribution: ContributionPlan;
  retirementContribution: ContributionPlan;
  expenseChanges: ExpenseChange[];
  socialSecurity: SocialSecurityPlan;
  includeHsaInRetirement: boolean;
  updatedAt: string;
};

export const createDefaultRetirementPlan = (): RetirementPlan => ({
  id: 'base-plan',
  name: 'Base Plan',
  currentAge: 41,
  retirementAge: 55,
  longevityAge: 105,
  annualRoiPercent: null,
  annualRetirementExpense: 0,
  taxableContribution: { monthlyAmount: 0, annualIncreasePercent: 0, endAge: 65 },
  retirementContribution: { monthlyAmount: 0, annualIncreasePercent: 0, endAge: 65 },
  expenseChanges: [],
  socialSecurity: { enabled: false, claimAge: 67, monthlyBenefit: 0, annualColaPercent: 0 },
  includeHsaInRetirement: false,
  updatedAt: '',
});
